### ISSUE TITLE
Tool to generate Facade from JSExports

### ISSUE THREAD

**Issue by jducoeur** [scala-js/scala-js#1703]
We are starting to have semi-frequent discussions along the lines of, "Can I modularize Scala.js within my application?  Can I put Scala.js libraries onto a CDN?"  This can't be done too naively -- due to Scala.js' full-program optimization, you can't just break it into parts, compile it as normal and get obvious results.

What you _can_ do, though, is to write "JavaScript" libraries in Scala.js -- exposing their APIs using JSExport, and treating them from the outside just like any other JavaScript library.  Something like that can be distributed via CDN and consumed by applications like a conventional JS library.  That's a pretty straightforward and practical way to write libraries or modules using SJS.

The only problem, currently, is code duplication.  When you do something like this, you need to write the library itself, using JSExport to define the entry points, _and_ you need to write a matching facade project, defining each of those entry points in terms of js.native.  This is fragile boilerplate, and likely to introduce inconsistencies and errors.

What would be very useful here is a tool that one can turn on using an sbt switch, that automatically generates a facade based on the JSExports -- basically a mirror that re-creates the Scala interface from the JavaScript view.  Applications would depend upon and use that generated facade, providing a reliable, strongly-typed connection between library and application.

My guess is that this requires a combination of an sbt module and a compiler plugin, which collects all of the JSExports from the library project, uses them to construct a facade project, and then compiles the facade project.  The facade project can then either be published to Maven (for public libraries) or depended upon by a higher-level project (for applications that want to factor out less-frequently-changing modules).

This should probably be marked as a Community project, seeking someone who wants to tackle it.

**Comment by mdedetrich** [comment]
A killer feature for this would be lazy loading of modules, which is pretty much required for SPA style websites when working with things like mobile devices (and general user experience with your web page)

Made a separate issue for that here https://github.com/scala-js/scala-js/issues/1751, but you essentially need interfaces to your modules so you know how to load the modules

**Comment by virtualirfan** [comment]
+1

**Comment by slorber** [comment]
Hey I''m not sure this is exactly the same topic but check this blog's comments as I share some concerns with you: http://underscore.io/blog/posts/2015/06/10/scalajs-scaladays.html

@mdedetrich personnaly I'm not needing lazy modules at all (as I embed my full app in a cordova shell, or I prefer to download once a big fat JS file than using multiple connections: mobile throughput is pretty nice nowadays!)

Just want to be able to easily split a big SPA app into multiple modules (some of them being potentially plain old JS modules).
Anyway once this is solved, you can still use Webpack or anything else to lazy load them :)

**Comment by mdedetrich** [comment]
The thing, stuff like Webpack uses `Common`/`AMD`/`require.js` etc etc, which means that either `SBT` needs to compile Javascript modules to follow that specification, or you need to manually write wrappers for all of your modules

We have the same issue, we are using `require.js` to do our module/injection system, but I am still evaluating whether or not its going to be possible to integrate well

**Comment by jducoeur** [comment]
I'm still going to push back, because the semantics matter.  sbt is irrelevant -- it's just a build tool.  Saying sbt needs to compile something is misleading -- sbt isn't the compiler, it's what *runs* the compiler.  And I don't see any reason why the main scalajs pipeline itself needs to be changed.  I believe someone needs to write *an sbt plugin* that *integrates* with the main pipeline, similar to the way @vmunier has done for Play integration.

That's an important distinction -- I think you need to demonstrate why it's better to change the core tools (which don't and shouldn't need to know anything about a specific framework like this) instead of building a new opt-in tool that works with them.  I don't think we disagree about what would be useful here, just where it belongs.  (Which matters, because there are very few people qualified to work in the core powertrain, but anybody can write a plugin...)

**Comment by siriux** [comment]
Any news on this, now that #2175 is included on the roadmap for 1.0?

**Comment by sjrd** [comment]
Why would #2175 being in the roadmap change anything about this issue? They are completely orthogonal concerns.

#2175 is in the roadmap because it requires language changes and support from core. This issue can be implemented outside of core.

**Comment by siriux** [comment]
Yes, I understand that, I didn't want to imply that any of them is a requirement for the other.

But once #2175 is solved, this issue will become much more useful and relevant. Because when you have a working module system, it seems obvious to use it from Scala.js code too, and for that you need facades.

So let me rephrase my question. Is there interest in reviving this issue, now that #2175 is in the roadmap for 1.0, and that makes automatic facades even more useful?

Cheers

**Comment by sjrd** [comment]
The issue *is* alive. But as the "community" label indicates, we do not intend to work on this in the core team. A solution should come from the community. With the manpower we have, we have to focus on things that need to be core.

**Comment by briantopping** [comment]
I just learned what the community label meant (thanks), and I've been doing OSS a lot longer than Github has been around. Not defending anyone, just pointing out that labels aren't always obvious.

**Comment by mkotsbak** [comment]
If such tool is made, it would be nice if it could also output Typescript definition files for the interface.

**Comment by david-bouyssie** [comment]
As @mkotsbak  said, it would be awesome to have something like https://github.com/miloszpp/scala-ts generalized to ScalaJS, in order to export Typescript definition files.

**Comment by sjrd** [comment]
I'm sympathetic to the needs exposed in this issue, but
* we're definitely not going to work on this ourselves, and
* it can and should be implemented in a separate repository, as a separate tool.

So I'm going to close this here. If someone wants to give it a shot, they are of course welcome to do so.
